File: schemas/auth-rbac/convex/oauth_providers.ts
// oauth_providers: Configuration for external OAuth/SSO providers your app authenticates against.
// This is the "consuming" side -- your app is the relying party.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const oauthProviders = defineTable({
  name: v.string(), // Display name (e.g., "Google", "Acme Corp SSO").
  slug: v.string(), // URL-safe identifier (e.g., "google", "github"). Used in callback URLs.
  strategy: v.string(), // "oauth2", "oidc", or "saml".
  clientId: v.string(), // OAuth client ID from the provider's developer console.
  clientSecret: v.optional(v.string()), // Encrypt at rest. Nullable for public clients using PKCE.
  authorizationUrl: v.optional(v.string()), // Override for custom/self-hosted providers.
  tokenUrl: v.optional(v.string()),
  userinfoUrl: v.optional(v.string()),
  scopes: v.array(v.string()), // Default scopes to request.
  enabled: v.boolean(), // Toggle on/off without deleting configuration.

  organizationId: v.optional(v.id("organizations")), // Org-scoped SSO. Null = available to all users.

  metadata: v.optional(v.any()), // Provider-specific config.
  updatedAt: v.number(),
})
  .index("by_slug", ["slug"])
  .index("by_organization_id", ["organizationId"])
  .index("by_enabled", ["enabled"]);

File: schemas/auth-rbac/spacetimedb/oauth_providers.rs
// oauth_providers: External OAuth/SSO provider configuration (Google, GitHub, corporate SAML, etc.).
// This is the "consuming" side — your app is the relying party.
// For when your app acts as an OAuth server, see oauth_clients instead.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = oauth_providers, public)]
pub struct OauthProvider {
    #[primary_key]
    pub id: String, // UUID

    pub name: String, // Display name shown in your UI (e.g., "Google", "Acme Corp SSO").

    #[unique]
    pub slug: String, // URL-safe identifier. Used in callback URLs (e.g., /auth/callback/google).

    pub strategy: String, // "oauth2", "oidc", or "saml". Determines which flow to use.

    pub client_id: String, // OAuth client ID from the provider's developer console.

    // Must be encrypted at rest. Nullable for public clients (mobile/SPA using PKCE without a secret).
    pub client_secret: Option<String>,

    pub authorization_url: Option<String>, // Override for custom/self-hosted providers.
    pub token_url: Option<String>,         // Override for custom providers.
    pub userinfo_url: Option<String>,      // Override for custom providers.

    pub scopes: Vec<String>, // Default scopes to request (e.g., ["openid", "profile", "email"]).

    #[index(btree)]
    pub enabled: bool, // Toggle a provider on/off without deleting its configuration.

    // Organization-scoped SSO: null = available to all users (e.g., "Sign in with Google").
    #[index(btree)]
    pub organization_id: Option<String>, // FK → organizations.id (cascade delete)

    pub metadata: Option<String>, // JSON. Provider-specific config that doesn't fit in standard fields.

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
